add Learn the concepts --> Peer-to-Peer Systems and Multi-Agent Systems (#262)
fetchai/docs: docs/learn_the_concepts/agent-based_and_multi-agents_systems.md

The _complexity of MASs_ can increase rapidly with the number of agents, the interactions among them, or the complexity of their behavior. System composition depends on the characteristics of individual agents and whether they are homogeneous or heterogeneous in physical and/or programming structure. Homogeneous agents are similar to each other or they are of the same type, whereas, heterogeneous agents are different and diverse in kind.

MAS have received tremendous attention from scholars in different disciplines, ranging from computer science to civil engineering, in order to solve complex problems by subdividing them into smaller tasks. The individual tasks are allocated to autonomous agents, and each one of them decides on a proper action to solve the task using multiple inputs (e.g. history of actions, interactions with its neighboring agents, and its goal).

> A good example of a MAS is represented by _smart-grids_, in which different agents work together collaboratively, but independently and in a self-interested manner. Another important application of such systems is found in _computer gaming_ and _computer simulations_. Such fields increasingly adopted MASs. Here, each participant has its own agent and interacts in the environment populated also by other agents. _Self-driving vehicles_ are another great example of an agent based system. In this respect, each vehicle acts as an agent, directed by their user and acting independently from other agents within a set of rules, such as not exceeding the speed limits, and responsive to the environment around them.

Reviewer: responsive to the environment around them -> being responsive to the environment around them.

As we can see, MAS have found multiple applications including _modeling complex systems_, _smart grids_ and _computer networks_. The main application domains of MAS are _ambient intelligence_, _grid computing_, _electronic business_, _the semantic web_, _bioinformatics and computational biology_, _monitoring and control_, _resource management_, _education_, _space_, _military_ and _manufacturing applications_, and so on.

> **Despite their wide applicability, there are still a number of challenges faced by MAS including coordination between agents, security, and task allocation.**

Reviewer: there are still a number of challenges faced by MAS including coordination between agents -> there are still several challenges faced by MAS including coordination between agents

# Agent-Based Systems

> An **agent** is a computer program that performs various actions _continuously_ and _autonomously_  (with limited or no interference) on behalf of an individual or an organization. Agents have behaviors, often described by simple rules, and interactions with other agents, which in turn influence their behaviors. The most important defining characteristic of an agent is its capability to act autonomously, that is, to act on its own without external direction in response to situations it encounters. Agents are endowed with behaviors that allow them to make independent decisions. Typically, agents are active, initiating their actions to achieve their internal goals, rather than merely passive, reactively responding to other agents and the environment. 

Reviewer 2: ```suggestion
An **agent** is a piece of software that _represents_ an entity (individual, organisation, or object) and performs various actions _continuously_ and _autonomously_  (with limited or no interference) on their behalf. 

The most important defining characteristic of an agent is its ability to act autonomously, that is, to act on its own without external direction in response to situations it encounters. Agents are endowed with behaviours that allow them to make independent decisions. 

Additionally, what differentiates agents with other software paradigms (e.g. smart contracts, web apps) is that as well as being reactive, responding to other agents and changes in their environments, agents are proactive, meaning they take the initiative and perform actions to achieve their goals.
```

Agents have **five key characteristics**:

> 1. **Representation**: The first key characteristic of agents is that they operate on behalf of an individual or organization. Agents are always operating for an entity. 

Reviewer 2: ```suggestion
1. **Representation**: agents are owned by and operate on behalf of an entity, for example an individual, family, company, government, or object, and look after their owner's interests.
```

> 2. **Autonomy**: Agents operate with limited or no interference, they do not need to be told what to do constantly but perform actions continuously according to the rules they have been given.

Reviewer 2: ```suggestion
2. **Autonomy**: Agents operate with limited or no interference, they do not need to be told what to do constantly but perform actions continuously according to their internal belief system.
```

> 3. **Self-interested**: They operate in their own interests and not necessarily in the interests of other agents.

Reviewer 2: ```suggestion
3. **Self-interested**: Each agent primarily looks after its own interests (which is aligned with those of its owner) and not necessarily the interests of other agents.
```

> 4. **Goal-oriented**:  agents have goals to achieve with respect to their behaviors. This allows an agent to compare the outcome of its behaviors relative to its goals and adjust its responses and behaviors in future interactions.

Reviewer 2: ```suggestion
4. **Goal-oriented**:  Agents have goals to achieve. This allows them to compare the outcome of their behaviours relative to their goals and adjust their responses and behaviours in future interactions.
```

> 5. **Reactive**: Agents are reactive. This is because they react to the circumstances around them and act consequently. A great example of this is an automatic thermostat, that turns the heating on or off when the temperature changes.

Reviewer 2: ```suggestion
5. **Reactive**: Agents respond to other agents, services, etc., and changes in their environments. A great example of this behaviour is an automatic heating system with a thermostat that constantly monitors the environment turns the heating on or off when the temperature changes.
```

> # Agent-Based Systems

Reviewer 2: ```suggestion
## Agent
```

> An **agent** is a piece of software that _represents_ an entity (individual, organisation, or object) and performs various actions _continuously_ and _autonomously_  (with limited or no interference) on their behalf. 

Reviewer 2: ```suggestion
An **agent** is a piece of software that _represents_ an entity (individual, organisation, or object) and acts _continuously_ and _autonomously_  (with limited or no interference) on their behalf. 
```

> The most important defining characteristic of an agent is its ability to act autonomously, that is, to act on its own without external direction in response to situations it encounters. Agents are endowed with behaviours that allow them to make independent decisions. 

Reviewer 2: ```suggestion
The most important defining characteristic of an agent is its autonomy; that is, to act on its own without external direction from its owner in response to situations it encounters. Agents are endowed with behaviours that allow them to make independent decisions. 
```

> What differentiates agents with other software paradigms (e.g. smart contracts, web apps) is that they are _reactive_, they respond to other agents and changes in their environments, and also _proactive_, meaning they take the initiative and perform actions to achieve their goals.

Reviewer 2: ```suggestion
What differentiates agents with other software paradigms (e.g. smart contracts, web apps) is that as well as being _reactive_, meaning they respond to other agents and changes in their environments, they are also _proactive_, which means that they take the initiative and perform actions to achieve their goals.
```

Agents have **five key characteristics**:

1. **Representation**: Agents are owned by and operate on behalf of an entity, for example an individual, family, company, government, or object, and look after their owner's interests. 

> 2. **Autonomy**: Agents operate with limited or no interference, they do not need to be told what to do constantly but perform actions continuously according to their internal reasoning system.

Reviewer 2: ```suggestion
2. **Autonomy**: Agents operate with limited or no interference and do not need to be constantly told what to do. They perform actions continuously according to their internal reasoning system.
```

3. **Self-interested**: Each agent primarily looks after its own interests (which is aligned with those of its owner) and not necessarily the interests of other agents.

4. **Proactive**:  Agents have goals to achieve allowing them to take the initiative and perform actions that get them closer to achieving their goals. This means an agent typically compares the outcome of different actions relative to its goals and selects the one that takes it closer to them.

> 5. **Reactive**: Agents respond to other agents, services, etc., and changes in their environments. A great example of this behaviour is a heating system with thermostat that constantly monitors the environment and turns the heating on or off when the temperature changes.

Reviewer 2: ```suggestion
5. **Reactive**: Agents respond to other agents, services, etc., and changes in their environments. A great example of this behaviour is a heating system with a thermostat that constantly monitors its environment and turns the heating on or off when the temperature changes.
```

> _Agent characteristics and behaviors may vary in their extent and sophistication_. An agent may differ from other agents, for instance, in the amount of information considered in the agent’s decisions process, the agent’s internal models of the external world, the agent’s view of the possible reactions of other agents in response to its actions, and the extent of memory of past events the agent retains and uses in making its decisions. Agents may also be endowed with  different amounts of resources or accumulate different levels of resources as a result of agent interactions, further differentiating agents.

Reviewer 2: ```suggestion
_Agent characteristics and behaviors may vary in their extent and sophistication_. An agent may differ from other agents, for instance, in the amount of information considered in the agent’s decisions process, its internal models of the external world, its view of the possible reactions of other agents in response to its actions, and the size of the memory of past events the agent retains and uses in making its decisions. Further differentiating factors are the amounts of resources used by agents or accumulated as a result of their interactions.
```

> Agent-based modeling has been used in an enormous variety of applications spanning the physical, biological, social, and management sciences. In the biological sciences, agent-based modeling is used to model cell behavior and interaction, the workings of the immune system, tissue growth, and disease processes. Archaeology and anthropology make use of large scale agent-based modeling by providing a virtual environment for long-vanished civilizations. In ecology, agent-based modeling is used to model diverse populations of individuals and their interactions (e.g. Agent-based epidemic and pandemic models). Agent-based models have also been developed in the fields of economics, sociology, anthropology, and cognitive science. Various social phenomena have been investigated using agent-based models that are not easily modeled using other approaches. Here, agent-based models are being used to analyze existing and hypothetical markets (e.g. agent-based simulation to model possible futures for a market directed towards space tourism, to analyze how companies represented by agents would compete and offer products to customers in this hypothetical market).

Reviewer 2: ```suggestion
## Agent-Based Modelling

Agent-based modelling has been used in enormous kinds of applications spanning the physical, biological, social, and management sciences. In biological sciences, agent-based modelling is used to simulate cell behavior and interaction, the workings of the immune system, tissue growth, and disease processes. Archaeology and anthropology make use of large scale agent-based modeling by providing a virtual environment for long-vanished civilizations. In ecology, agent-based modeling is used to model diverse populations of individuals and their interactions (e.g. Agent-based epidemic and pandemic models). 

Agent-based models have also been developed in the fields of economics, sociology, anthropology, and cognitive science. Various social phenomena have been investigated using agent-based models that are not easily modeled using other approaches. Here, agent-based models are being used to analyze existing and hypothetical markets. For example, agent-based simulation is is used to model possible futures for a market that is directed towards space tourism, to analyze how companies represented by agents would compete and offer products to customers in this hypothetical market.
```

Reviewer 2: Add a paragraph at the beginning of agent-based modelling introducing introducing what it is.

> As it is possible to think, there are a growing number of agent-based applications in a variety of fields and disciplines directed towards the study of a large number of real-world applications. Examples of agent-based systems, for which large-scale agent-based models have been developed, include: traffic, air traffic control, military command and control, physical infrastructures and markets, such as electric power and integrated energy markets. 

Reviewer 2: ```suggestion
Unsurprisingly, there are a growing number of agent-based applications in a variety of fields and disciplines that focus on the study of large numbers of real-world phenomena. Some other examples of systems, for which large-scale agent-based models have been developed, include: traffic, air traffic control, military command and control, physical infrastructures and markets such as electric power and integrated energy markets. 
```

> # Multi-Agents Systems

Reviewer 2: ```suggestion
## Multi-Agents Systems
```

Agent-based modelling has been used in enormous kinds of applications spanning the physical, biological, social, and management sciences. In biological sciences, agent-based modelling is used to simulate cell behavior and interaction, the workings of the immune system, tissue growth, and disease processes. Archaeology and anthropology make use of large scale agent-based modeling by providing a virtual environment for long-vanished civilizations. In ecology, agent-based modeling is used to model diverse populations of individuals and their interactions (e.g. Agent-based epidemic and pandemic models). 

> Agent-based models have also been developed in the fields of economics, sociology, anthropology, and cognitive science. Various social phenomena have been investigated using agent-based models that are not easily modeled using other approaches. Here, agent-based models are being used to analyze existing and hypothetical markets. For example, agent-based simulation is is used to model possible futures for a market that is directed towards space tourism, to analyze how companies represented by agents would compete and offer products to customers in this hypothetical market.

Reviewer: For example, agent-based simulation is is used to model possible futures for a market -> For example, agent-based simulation is used to model possible futures for a market.
